### ISSUE TITLE
504 Gateway Time-out. Search for products in backend

### ISSUE THREAD

**Issue by steffanhalv** [woocommerce/woocommerce#20742]
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. Add 60K + products to your store
2. Try search in backend, ex.: https://example.com/wp-admin/edit.php?s=107646-00211-117&post_status=all&post_type=product&action=-1&product_cat&product_type&stock_status&paged=1&action2=-1
3. Request takes VERY long time

## Expected/actual behavior

Be super fast, like frontend. Ex. this works just fine:
https://example.com/?s=107646-00211-117

Also search for products using rest api is much faster and works just fine.

## Isolating the problem

- [x] This bug happens with only WooCommerce plugin active
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [x] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>
```
`
### WordPress Environment ###

Home URL: https://haagen.no
Site URL: https://haagen.no
WC Version: 3.4.3
Log Directory Writable: ✔
WP Version: 4.9.7
WP Multisite: –
WP Memory Limit: 4 GB
WP Debug Mode: –
WP Cron: ✔
Language: nb_NO
External object cache: –

### Server Environment ###

Server Info: nginx/1.10.3
PHP Version: 7.0.30-0+deb9u1 - Vi anbefaler å bruke PHP versjon 7.2 eller nyere for større ytelse og sikkerhet. Hvordan oppgradere PHP til nyere versjon
PHP Post Max Size: 240 MB
PHP Time Limit: 30024
PHP Max Input Vars: 1000
cURL Version: 7.52.1
OpenSSL/1.0.2l

SUHOSIN Installed: –
MySQL Version: 5.7.14-google-log
Max Upload Size: 240 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.4.3
WC Database Prefix: hg_
MaxMind GeoIP Database: ✔
Total database størrelse: 985.37MB
Database data størrelse: 731.48MB
Database indeks størrelse: 253.89MB
hg_woocommerce_sessions: Data: 4.02MB + Indeks: 0.06MB
hg_woocommerce_api_keys: Data: 0.02MB + Indeks: 0.03MB
hg_woocommerce_attribute_taxonomies: Data: 0.02MB + Indeks: 0.02MB
hg_woocommerce_downloadable_product_permissions: Data: 0.02MB + Indeks: 0.05MB
hg_woocommerce_order_items: Data: 0.20MB + Indeks: 0.13MB
hg_woocommerce_order_itemmeta: Data: 1.52MB + Indeks: 2.00MB
hg_woocommerce_tax_rates: Data: 0.02MB + Indeks: 0.06MB
hg_woocommerce_tax_rate_locations: Data: 0.02MB + Indeks: 0.03MB
hg_woocommerce_shipping_zones: Data: 0.02MB + Indeks: 0.00MB
hg_woocommerce_shipping_zone_locations: Data: 0.02MB + Indeks: 0.03MB
hg_woocommerce_shipping_zone_methods: Data: 0.02MB + Indeks: 0.00MB
hg_woocommerce_payment_tokens: Data: 0.02MB + Indeks: 0.02MB
hg_woocommerce_payment_tokenmeta: Data: 0.02MB + Indeks: 0.03MB
hg_woocommerce_log: Data: 0.02MB + Indeks: 0.02MB
hg_commentmeta: Data: 0.02MB + Indeks: 0.03MB
hg_comments: Data: 1.52MB + Indeks: 1.20MB
hg_links: Data: 0.02MB + Indeks: 0.02MB
hg_options: Data: 2.23MB + Indeks: 1.02MB
hg_postmeta: Data: 234.81MB + Indeks: 173.30MB
hg_posts: Data: 475.97MB + Indeks: 67.00MB
hg_social_users: Data: 0.02MB + Indeks: 0.02MB
hg_termmeta: Data: 0.02MB + Indeks: 0.03MB
hg_terms: Data: 0.02MB + Indeks: 0.03MB
hg_term_relationships: Data: 5.52MB + Indeks: 3.52MB
hg_term_taxonomy: Data: 0.02MB + Indeks: 0.03MB
hg_usermeta: Data: 3.52MB + Indeks: 4.03MB
hg_users: Data: 0.23MB + Indeks: 0.23MB
hg_wc_download_log: Data: 0.02MB + Indeks: 0.03MB
hg_wc_webhooks: Data: 0.02MB + Indeks: 0.02MB
hg_woocommerce_heidelpaycw_customer_contexts: Data: 0.02MB + Indeks: 0.02MB
hg_woocommerce_heidelpaycw_documents: Data: 0.02MB + Indeks: 0.02MB
hg_woocommerce_heidelpaycw_ecc: Data: 0.09MB + Indeks: 0.00MB
hg_woocommerce_heidelpaycw_schema_version: Data: 0.02MB + Indeks: 0.02MB
hg_woocommerce_heidelpaycw_storage: Data: 0.02MB + Indeks: 0.00MB
hg_woocommerce_heidelpaycw_transactions: Data: 0.36MB + Indeks: 0.00MB
hg_wpio_queue: Data: 0.02MB + Indeks: 0.00MB
wp_commentmeta: Data: 0.02MB + Indeks: 0.03MB
wp_comments: Data: 0.02MB + Indeks: 0.08MB
wp_links: Data: 0.02MB + Indeks: 0.02MB
wp_options: Data: 0.38MB + Indeks: 0.02MB
wp_postmeta: Data: 0.08MB + Indeks: 0.03MB
wp_posts: Data: 0.05MB + Indeks: 0.06MB
wp_termmeta: Data: 0.02MB + Indeks: 0.03MB
wp_terms: Data: 0.02MB + Indeks: 0.03MB
wp_term_relationships: Data: 0.02MB + Indeks: 0.02MB
wp_term_taxonomy: Data: 0.02MB + Indeks: 0.03MB
wp_usermeta: Data: 0.02MB + Indeks: 0.03MB
wp_users: Data: 0.02MB + Indeks: 0.05MB
wp_wc_download_log: Data: 0.02MB + Indeks: 0.03MB
wp_wc_webhooks: Data: 0.02MB + Indeks: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Indeks: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Indeks: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Indeks: 0.05MB
wp_woocommerce_log: Data: 0.02MB + Indeks: 0.02MB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Indeks: 0.03MB
wp_woocommerce_order_items: Data: 0.02MB + Indeks: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Indeks: 0.03MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Indeks: 0.02MB
wp_woocommerce_sessions: Data: 0.02MB + Indeks: 0.02MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Indeks: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Indeks: 0.03MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Indeks: 0.00MB
wp_woocommerce_tax_rates: Data: 0.02MB + Indeks: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Indeks: 0.03MB

### Post Type Counts ###

amn_exact-metrics: 1
attachment: 262274
blocks: 1
custom_css: 2
customize_changeset: 11
jetpack_migration: 2
jp_img_sitemap: 32
jp_img_sitemap_index: 1
jp_sitemap: 1
jp_sitemap_master: 1
nav_menu_item: 20
page: 27
post: 3
product: 71761
question_answer: 4
revision: 107
shop_coupon: 1
shop_order: 4183
wpcf7_contact_form: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (20) ###

Akismet Anti-Spam: av Automattic – 4.0.8
Google reCAPTCHA by Astoundify: av Astoundify – 1.0.1
Better Search Replace: av Delicious Brains – 1.3.2
Contact Form 7: av Takayuki Miyoshi – 5.0.2
Placeholder Image for Woocommerce: av Игорь Тронь – 1.0.3 – Ikke testet med den aktive versjonen av WooCommerce
Google Tag Manager for Wordpress: av Thomas Geiger – 1.8.1
Haagen: av Siglar Development AS – 1.0.0
Import users from CSV with meta: av codection – 1.11.3.5
Loco Translate: av Tim Whitlock – 2.1.4
SendGrid: av SendGrid – 1.11.8
StorePilot: av StorePilot AS – 1.0.0
User Role Editor: av Vladimir Garagulya – 4.44
Vipps for WooCommerce: av Siglar Development AS – 1.4.2 – Ikke testet med den aktive versjonen av WooCommerce
Woo Email Control: av Ian Young – 1.061 – Ikke testet med den aktive versjonen av WooCommerce
WooCommerce: av Automattic – 3.4.3
WooCommerce HeidelpayCw: av customweb GmbH – 3.0.185 – Ikke testet med den aktive versjonen av WooCommerce
WooDiscuz - WooCommerce Comments: av gVectors Team (A. Chakhoyan
G. Zakaryan
H. Martirosyan) – 2.2.1 – Ikke testet med den aktive versjonen av WooCommerce

WooSidebars: av WooCommerce – 1.4.5
WordPress Importer: av wordpressdotorg – 0.6.4
WP File Manager: av mndpsingh287 – 2.8

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: NOK (kr)
Currency Position: left_space
Thousand Separator: .
Decimal Separator: .
Number of Decimals: 0
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Butikkbase: #216 - /butikk/
Handlekurv: #215 - /handlekurv/
Til kassen: #214 - /kasse/
Min konto: #213 - /min-konto/
Vilkår og betingelser: #39746 - /vilkar-og-betingelser/

### Theme ###

Name: Haagen
Version: 2.2.8
Author URL: https://woocommerce.com/
Child Theme: ❌ – Hvis du endrer WooCommerce eller et foreldertema du ikke selv har bygget
anbefaler vi å bruke et dattertema. Se: Hvordan lage et dattertema

WooCommerce Support: ✔

### Templates ###

Overrides: haagen/woocommerce/archive-product.php versjon - er utdatert. Kjerneversjonen er 3.4.0
haagen/woocommerce/cart/cart.php versjon 3.3.0 er utdatert. Kjerneversjonen er 3.4.0
haagen/woocommerce/checkout/form-billing.php
haagen/woocommerce/checkout/form-checkout.php
haagen/woocommerce/checkout/form-shipping.php
haagen/woocommerce/checkout/payment-method.php
haagen/woocommerce/checkout/thankyou.php
haagen/woocommerce/content-single-product.php versjon 3.0.0 er utdatert. Kjerneversjonen er 3.4.0
haagen/woocommerce/order/order-details-customer.php
haagen/woocommerce/order/order-details.php

Outdated Templates: ❌
					
					
						Lær hvordan å oppdatere
`
```
</details>

**Comment by claudiosanches** [comment]
@rodrigoprimo do you want to check if there is something that we can do to improve before search in admin before going to custom tables for products?

**Comment by rodrigoprimo** [comment]
IMHO this is a duplicate of #18325 so I'm closing it. @steffanhalv, please reopen if I missed something. We are currently working on a plugin to use a custom table instead of wp_postmeta to store product data. While this is not ready, I would suggest you add a custom index to wp_postmeta (assuming the problem you have is with the SKU part of the query to search for products):

```
ALTER TABLE wp_postmeta ADD INDEX YOUR_CUSTOM_PREFIX_meta_key_value (meta_key(191), meta_value(100))
```

Please make sure to replace YOUR_CUSTOM_PREFIX with a unique prefix.

You could use `query-monitor` to check the time of the queries that are executed while searching for products in the admin to check if my assumption is correct and to confirm if the index helps.
